src/storage/src/hummock/cache.rs

```diff
@@ -0,0 +1,664 @@
+            strict_capacity_limit,
+            lru,
+            table: LruHandleTable::new(),
+        }
+    }
+
+    unsafe fn lru_remove(&mut self, e: *mut LruHandle<K, T>) {
+        (*(*e).next).prev = (*e).prev;
+        (*(*e).prev).next = (*e).next;
+        (*e).prev = null_mut();
+        (*e).next = null_mut();
+        self.lru_usage -= (*e).charge;
+    }
+
+    // insert entry in the end of the linked-list
+    unsafe fn lru_insert(&mut self, e: *mut LruHandle<K, T>) {
+        (*e).next = self.lru.as_mut();
+        (*e).prev = self.lru.prev;
+        (*(*e).prev).next = e;
+        (*(*e).next).prev = e;
+        self.lru_usage += (*e).charge;
+    }
+
+    unsafe fn evict_from_lru(&mut self, charge: usize, last_reference_list: &mut Vec<T>) {
+        while self.usage + charge > self.capacity && !std::ptr::eq(self.lru.next, self.lru.as_mut())
+        {
+            let old_ptr = self.lru.next;
+            self.table.remove((*old_ptr).hash, &(*old_ptr).key);
+            assert!((*old_ptr).is_in_cache());
+            self.lru_remove(old_ptr);
+            (*old_ptr).set_in_cache(false);
+            self.usage -= (*old_ptr).charge;
+            let mut node = Box::from_raw(old_ptr);
+            let data = node.value.take().unwrap();
+            last_reference_list.push(data);
+            self.recyle_handle_object(node);
+        }
+    }
+
+    fn recyle_handle_object(&mut self, mut node: Box<LruHandle<K, T>>) {
+        if self.object_pool.len() < self.object_pool.capacity() {
+            node.next_hash = null_mut();
+            node.next = null_mut();
+            node.prev = null_mut();
+            self.object_pool.push(node);
+        }
+    }
+
+    unsafe fn insert(
+        &mut self,
+        key: K,
+        hash: u64,
+        charge: usize,
+        value: T,
+        last_reference_list: &mut Vec<T>,
+    ) -> *mut LruHandle<K, T> {
+        let mut handle = if let Some(mut h) = self.object_pool.pop() {
+            h.key = key;
+            h.value = Some(value);
+            h
+        } else {
+            Box::new(LruHandle::new(key, Some(value)))
+        };
+        handle.charge = charge;
+        handle.hash = hash;
+        handle.refs = 0;
+        handle.flags = 0;
+        handle.set_in_cache(true);
+        self.evict_from_lru(charge, last_reference_list);
+        if self.usage + charge > self.capacity && self.strict_capacity_limit {
+            handle.set_in_cache(false);
+            let data = handle.value.take().unwrap();
+            last_reference_list.push(data);
+            self.recyle_handle_object(handle);
+            null_mut()
+        } else {
+            let ptr = Box::into_raw(handle);
+            let old = self.table.insert(hash, ptr);
+            if !old.is_null() {
+                let data = self.remove_cache_handle(old).unwrap();
+                last_reference_list.push(data);
+            }
+            self.usage += charge;
+            (*ptr).add_ref();
```

Comment:
We only `add_ref()` once when the handle is inserted but not `add_ref()` for the returned handle in L413. Will this be a problem?

Reply:
Emmm, why? We only need `add_ref()` once.